HeavyFlavorAnalysis/RecoDecay/src/BPHKinematicFit.cc

```diff
@@ -0,0 +1,431 @@
+  if ( allParticles.size() != daughFull().size() ) return kinTree;
+  vector<RefCountedKinematicParticle> kComp;
+  vector<RefCountedKinematicParticle> kTail;
+  if ( name != "" ) {
+    const BPHRecoCandidate* comp = getComp( name ).get();
+    if ( comp == 0 ) {
+      edm::LogPrint( "ParticleNotFound" )
+                  << "BPHKinematicFit::kinematicTree: "
+                  << name << " daughter not found";
+      return kinTree;
+    }
+    const vector<string>& names = comp->daugNames();
+    int ns;
+    int nn = ns = names.size();
+    vector<string> nfull( nn + 1 );
+    nfull[nn] = "*";
+    while ( nn-- ) nfull[nn] = name + "/" + names[nn];
+    vector<RefCountedKinematicParticle> kPart = kinParticles( nfull );
+    vector<RefCountedKinematicParticle>::const_iterator iter = kPart.begin();
+    vector<RefCountedKinematicParticle>::const_iterator imid = iter + ns;
+    vector<RefCountedKinematicParticle>::const_iterator iend = kPart.end();
+    kComp.insert( kComp.end(), iter, imid );
+    kTail.insert( kTail.end(), imid, iend );
+  }
+  else {
+    kComp = allParticles;
+  }
+  try {
+    KinematicParticleVertexFitter vtxFitter;
+    RefCountedKinematicTree compTree = vtxFitter.fit( kComp );
+    if ( compTree->isEmpty() ) return kinTree;
+    KinematicParticleFitter kinFitter;
+    compTree = kinFitter.fit( kc, compTree );
+    if ( compTree->isEmpty() ) return kinTree;
+    compTree->movePointerToTheTop();
+    if ( kTail.size() ) {
+      RefCountedKinematicParticle compPart = compTree->currentParticle();
+      if ( !compPart->currentState().isValid() ) return kinTree;
+      kTail.push_back( compPart );
+      kinTree = vtxFitter.fit( kTail );
+    }
+    else {
+      return ( kinTree = compTree );
+    }
+  }
+  catch ( std::exception e ) {
+    edm::LogPrint( "FitFailed" )
+                << "BPHKinematicFit::kinematicTree: "
+                << "kin fit reset";
+    kinTree = RefCountedKinematicTree( 0 );
+  }
+  return kinTree;
+}
+
+
+const RefCountedKinematicTree& BPHKinematicFit::kinematicTree(
+                               const string& name,
+                               MultiTrackKinematicConstraint* kc ) const {
+  kinTree = RefCountedKinematicTree( 0 );
+  updatedFit = true;
+  kinParticles();
+  if ( allParticles.size() != daughFull().size() ) return kinTree;
+  vector<string> nfull;
+  if ( name != "" ) {
+    const BPHRecoCandidate* comp = getComp( name ).get();
+    if ( comp == 0 ) {
+      edm::LogPrint( "ParticleNotFound" )
+                  << "BPHKinematicFit::kinematicTree: "
+                  << name << " daughter not found";
+      return kinTree;
+    }
+    const vector<string>& names = comp->daugNames();
+    int nn = names.size();
+    nfull.resize( nn + 1 );
+    nfull[nn] = "*";
+    while ( nn-- ) nfull[nn] = name + "/" + names[nn];
+  }
+  else {
+    nfull.push_back( "*" );
+  }
+  try {
+    KinematicConstrainedVertexFitter cvf;
+    kinTree = cvf.fit( kinParticles( nfull ), kc );
```

Comment:
when does the fit throw?
We should not use exceptions for runtime error message exchanges